src/measure_theory/bochner_integration.lean

```diff
@@ -16,32 +45,537 @@ set_option class.instance_max_depth 100
+def to_simple_func (f : α →ₛ β) : measure_theory.simple_func α β := classical.some f.2
+
+/-- `f.to_simple_func` is measurable. -/
+protected lemma measurable (f : α →ₛ β) : measurable f.to_simple_func := f.to_simple_func.measurable
+
+/-- `f.to_simple_func` is integrable. -/
+protected lemma integrable (f : α →ₛ β) : integrable f.to_simple_func :=
+let ⟨h, _⟩ := classical.some_spec f.2 in h
+
+lemma to_simple_func_to_fun (f : α →ₛ β) : ∀ₘ a, (f.to_simple_func) a = f.1.to_fun a :=
+begin
+  rw ← l1.mk_eq_mk (f.to_simple_func) f.1.to_fun f.measurable f.integrable
+    f.1.measurable f.1.integrable,
+  rw ← self_eq_mk f.1,
+  rcases classical.some_spec f.2 with ⟨_, h⟩,
+  exact h
+end
+
+lemma mk_eq_mk' (f g : α →ₛ β) (hf hg) :
+  mk f.to_simple_func hf = mk g.to_simple_func hg ↔ (∀ₘ a, f.to_simple_func a = g.to_simple_func a) :=
+by { rw mk_eq_mk }
+
+lemma ext_iff {f g : α →ₛ β} : f = g ↔ (∀ₘ a, f.to_simple_func a = g.to_simple_func a) :=
+begin
+  rcases f with ⟨f, hf⟩, rcases g with ⟨g, hg⟩,
+  rcases classical.some_spec hf with ⟨hfi, hf⟩,
+  rcases classical.some_spec hg with ⟨hfi, hg⟩,
+  rw [subtype.mk_eq_mk, l1.ext_iff],
+  repeat { assumption }
+end
+
+lemma mk_to_simple_func (f : α →ₛ β) : mk f.to_simple_func f.integrable = f :=
+begin
+  rcases f with ⟨f, hf⟩,
+  rcases classical.some_spec hf with ⟨_, hf⟩,
+  simp only [mk, subtype.mk_eq_mk],
+  exact hf
+end
+
+lemma to_simple_func_mk (f : measure_theory.simple_func α β) (hf) :
+  ∀ₘ a, to_simple_func (mk f hf) a = f a :=
+let f' := to_simple_func (mk f hf) in
+have h : mk f' (mk f hf).integrable = mk f hf := mk_to_simple_func _,
+by { rw mk_eq_mk at h, exact h }
+
+section instances
+
+instance : has_zero (α →ₛ β) := ⟨⟨0, ⟨0, ⟨integrable_zero, mk_zero α β⟩⟩⟩⟩
+
+variables (α β)
+
+lemma zero_def : (0 : α →ₛ β) = ⟨0, ⟨0, ⟨integrable_zero, mk_zero α β⟩⟩⟩ := rfl
+
+lemma mk_zero : mk (0 : measure_theory.simple_func α β) integrable_zero = 0 := rfl
 
-instance :  has_coe (α →ₛ γ) (α →₁ γ) := ⟨subtype.val⟩
+lemma zero_to_simple_func : ∀ₘ a, (0 : α →ₛ β).to_simple_func a = 0 :=
+begin
+  filter_upwards [to_simple_func_to_fun (0 : α →ₛ β), l1.zero_to_fun α β],
+  assume a,
+  simp only [mem_set_of_eq],
+  assume h,
+  rw h,
+  assume h,
+  exact h
+end
+
+variables {α β}
+
+instance : has_add (α →ₛ β) := ⟨λ f g,
+  let f' := f.to_simple_func in
+  let g' := g.to_simple_func in
+  have hfm : measurable f' := f'.measurable,
+  have hgm : measurable g' := g'.measurable,
+  have hfi : integrable f' := f.integrable,
+  have hgi : integrable g' := g.integrable,
+  ⟨f.1 + g.1, ⟨f' + g', ⟨integrable_add hfm hgm hfi hgi,
+    begin
+      rcases f with ⟨f, hf⟩, rcases g with ⟨g, hg⟩,
+      rcases classical.some_spec hf with ⟨_, hf⟩,
+      rcases classical.some_spec hg with ⟨_, hg⟩,
+      have : l1.mk f' hfm hfi + l1.mk g' hgm hgi = f + g, { rw [←hf, ←hg], refl },
+      rw ← mk_add at this, exact this
+    end⟩⟩⟩⟩
+
+lemma add_def (f g : α →ₛ β) : f + g =
```

Comment:
Do you really need this? Typically, `simp [add]` (or `dsimp [add]` if this doesn't work) will unfold the definition of `add`.

Reply:
I'm not sure. I feel they typically do not understand what I want. 

For example : 
Goal is `-⟨f, hf⟩ + ⟨f, hf⟩ = 0`. I want to rewrite this to get `-f + f = 0`. Now I do the following:
Do
```lean
simp [has_zero.zero, has_add.add, has_neg.neg]
```
Get
```lean
add_semigroup.add ⟨add_group.neg f, _⟩ ⟨f, hf⟩ = add_monoid.zero (α →₁ₛ β)
```
Do
```lean
simp [add_semigroup.add]
```
Get
```lean
add_monoid.add ⟨add_group.neg f, _⟩ ⟨f, hf⟩ = add_monoid.zero (α →₁ₛ β)
```
Do
```lean
 simp [add_monoid.add]
```
Get
```lean
⟨add_group.neg f, _⟩ + ⟨f, hf⟩ = add_monoid.zero (α →₁ₛ β)
```
Do
```lean
simp [has_add.add]
```
Get
```lean
⟨add_semigroup.add (add_group.neg f) f, _⟩ = add_monoid.zero (α →₁ₛ β)
```
So you get the idea.